<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

genomics-bigquery 1,000 Genomes
=================

### Additional Resources
* Schema
* [Provenance](./provenance)
* [Data Stories](./data-stories) such as
 * [Exploring the phenotypic data](./data-stories/exploring-the-phenotypic-data)
 * [Understanding Alternate Alleles in 1,000 Genomes](./data-stories/understanding-alternate-alleles)
 * [Literate Programming with R and BigQuery](./data-stories/literate-programming-demo)
* [Index of variant analyses](./analyses-catalog.md)

### Diving right in

The following query returns the proportion of variants that have been reported in the [dbSNP database](http://www.ncbi.nlm.nih.gov/projects/SNP/snp_summary.cgi?build_id=132) [version 132](http://www.1000genomes.org/category/variants), by chromosome, in the dataset:

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
require(bigrquery)
sql = readChar('./sql/ratio-of-dbsnp-variants-by-chromosome.sql',
               nchars=1e6)
cat(sql)
billing_project = "google.com:biggene" # put your projectID here
result = query_exec(project="google.com:biggene", dataset="1000genomes", query=sql, billing = billing_project)
```

We see the tabular results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
require(xtable)
print(xtable(result,digits=6), type = "html", include.rownames = F)
```

And visually:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align='center'}
require(ggplot2)
qplot(total_num_entries, variants_num_variants, label=variants_contig, data=result) + ylab("Number of dbSNP Variants") + xlab("Number of Variants") + ggtitle("dbSNP Variant Count vs. Total Variant Count by Chromosome") + geom_text(aes(label=variants_contig),hjust=-1, vjust=0)
```
