genomics-bigquery 1,000 Genomes
=================

### Additional Resources
* Schema
* [Provenance](./provenance)
* [Data Stories](./data-stories)
 * [Exploring the phenotypic data](./data-stories/exploring-the-phenotypic-data)
 * [Understanding Alternate Alleles in 1,000 Genomes](./data-stories/understanding-alternate-alleles)
 * [Reproducing the output of vcfstats](./data-stories/reproducing-vcfstats)
 * [Examining the clinical significance of variants](./data-stories/examining-clinical-significance)
 * [Literate Programming with R and BigQuery](./data-stories/literate-programming-demo)
* Index of variant analyses

### Diving right in

The following query returns the proportion of variants that have been reported in the [dbSNP database](http://www.ncbi.nlm.nih.gov/projects/SNP/snp_summary.cgi?build_id=132) [version 132](http://www.1000genomes.org/category/variants), by chromosome, in the dataset:

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
require(bigrquery)
sql = readChar('./sql/ratio-of-dbsnp-variants-by-chromosome.sql',
               nchars=1e6)
cat(sql)
billing_project = "google.com:biggene" # put your projectID here
result = query_exec(project="google.com:biggene", dataset="1000genomes", query=sql, billing = billing_project)
```

We see the tabular results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
require(xtable)
print(xtable(result,digits=6), type = "html", include.rownames = F)
```

And visually:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align='center'}
require(ggplot2)
qplot(total_num_entries, variants_num_variants, label=variants_contig, data=result) + ylab("Number of dbSNP Variants") + xlab("Number of Variants") + ggtitle("dbSNP Variant Count vs. Total Variant Count by Chromosome") + geom_text(aes(label=variants_contig),hjust=-1, vjust=0)
```
