<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2014 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

genomics-bigquery 1,000 Genomes
=================

### Additional Resources
* [Schema](https://bigquery.cloud.google.com/table/google.com:biggene:1000genomes.variants1kG?pli=1)
* [Provenance](./provenance)
* [Data Stories](./data-stories) such as
 * [Exploring the phenotypic data](./data-stories/exploring-the-phenotypic-data)
 * [Understanding Alternate Alleles in 1,000 Genomes](./data-stories/understanding-alternate-alleles)
 * [Literate Programming with R and BigQuery](./data-stories/literate-programming-demo)
* [Index of variant analyses](./sql)

### Diving right in
The following query returns the proportion of variants that have been reported in the [dbSNP database](http://www.ncbi.nlm.nih.gov/projects/SNP/snp_summary.cgi?build_id=132) [version 132](http://www.1000genomes.org/category/variants), by chromosome, across the entirety of the 1,000 Genomes low coverage variant data for 1,092 individuals:

```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
require(bigrquery)
require(ggplot2)
require(xtable)
require(testthat)
billing_project = 'google.com:biggene' # put your projectID here
sql = readChar('./sql/ratio-of-dbsnp-variants-by-chromosome.sql',
               nchars=1e6)
cat(sql)
result = query_exec(project="google.com:biggene", dataset="1000genomes",
                    query=sql, billing = billing_project)
```

We see the tabular results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
print(xtable(result,digits=6), type = "html", include.rownames = F)
```

And visually:
```{r dbSNP Variants, echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align='center', fig.width=12, fig.height=8}
qplot(num_variants, num_dbsnp_variants, color=num_variants, data=result) + 
  scale_colour_gradient("Number of Variants", labels = function(x)round(x)) +
  ylab("Number of dbSNP Variants") + 
  xlab("Number of Variants") + 
  ggtitle("dbSNP Variant Count vs. Total Variant Count by Chromosome") + 
  geom_text(aes(label=contig), hjust=-1, vjust=0)
```

## Variant Metadata
The 1000 Genomes variant data is stored in the [variants1kG](https://bigquery.cloud.google.com/table/google.com:biggene:1000genomes.variants1kG?pli=1) table.  Every record in the variants table maps to a single site (line) in the [VCF](http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41) file.  See the [schema](https://bigquery.cloud.google.com/table/google.com:biggene:1000genomes.variants1kG?pli=1) for more detail.

Show variants within BRCA1:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
sql = readChar('./sql/variant-level-data-for-brca1.sql',
               nchars=1e6)
cat(sql)
result = query_exec(project="google.com:biggene", dataset="1000genomes",
                    query=sql, billing = billing_project)
```
Number of rows returned by this query:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
cat(nrow(result))
```
Examing the first few rows, we see:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
print(xtable(head(result)), type = "html", include.rownames = F)
```
One can add more columns to the SELECT statement corresponding to INFO fields of interest as desired.

## Sample Data
Show variants for a paricular sample within BRCA1:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
sql = readChar('./sql/sample-level-data-for-brca1.sql',
               nchars=1e6)
cat(sql)
result = query_exec(project="google.com:biggene", dataset="1000genomes",
                    query=sql, billing = billing_project)
```
Number of rows returned by this query:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
cat(nrow(result))
```
Examing the first few rows, we see:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
print(xtable(head(result)), type = "html", include.rownames = F)
```
Note that this is equivalent to the [vcf-query](http://vcftools.sourceforge.net/perl_module.html#vcf-query) command
```
vcf-query ALL.chr17.phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf.gz 17:41196312-41277500 -c HG00100
```
Lastly, let's get an overview of how much variation is shared across the samples.
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
sql = readChar('./sql/shared-variant-counts.sql',
               nchars=1e6)
cat(sql)
result = query_exec(project="google.com:biggene", dataset="1000genomes",
                    query=sql, billing = billing_project)
```
Number of rows returned by this query:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
cat(nrow(result))
```
Examing the first few rows, we see that a substantial number of variants are shared by **none** of the samples but a larger number of the variants are shared by only one sample:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
print(xtable(head(result)), type = "html", include.rownames = F)
```
Looking at the last few rows in the result, we see that some variants are shared by all samples:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results='asis'}
print(xtable(tail(result)), type = "html", include.rownames = F)
```
And visually:
```{r shared Variants, echo=FALSE, message=FALSE, warning=FALSE, comment=NA, fig.align='center', fig.width=12, fig.height=8}
ggplot(result, aes(x=num_samples_with_variant, 
                  y=num_variants_shared_by_this_many_samples,
                  color=num_variants_shared_by_this_many_samples)) +
  geom_point() +
  scale_y_log10() +
  scale_colour_gradient("Number of Variants", trans = "log", labels = function(x)round(x)) +
  ylab("Number of variants shared by this many samples") +
  xlab("Number of samples sharing the variants") +
  ggtitle("An Overview of How Variation is Shared Across Samples") 
```
